### PR TITLE
[SecurityBundle] UserPasswordEncoderCommand: Improve & simplify the command usage

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #13988 |
| License | MIT |

Overlaps #14017 (might replace or follow it) . Sorry if it is considered as a duplicate, but the debate has evolved, and I think the arguments ordering isn't the best nor single way to improve this command usage anymore.
Thank you @saro0h for having considered the mentioned issues and spent time on it.
# Salt option & salt generation

Thanks to @ircmaxell & @inanimatt, we came to the conclusion that the use-cases for the `salt` option/argument are pretty arguable. So I suggest to get rid of it: **a salt will always be generated by the command.**
### Generated salt

The generated salt is now in the Table output:
`security:encode-password test -n`
screenshot 2015-03-30 a 21 38 21

The "Generated salt" row and the last comment about the salt aren't present if the new `empty-salt` option is provided (see below).
# New empty-salt option

As some encoders might generate their own built-in salts (like the `BCryptPasswordEncoder`) and some custom encoders could do the same (or not require a salt at all), I suggest a new option: `empty-salt`.
This option will not provide any salt to the configured encoder, which will generate its own, then.

With the interactive way, the user will always be asked confirmation for the salt generation if the `empty-salt` option isn't set:

`security:encode-password password`
screenshot 2015-03-30 a 21 38 43
### bcrypt encoder

As the `BCryptPasswordEncoder` is shipped with the security component, and listening to @inanimatt valuable comments in #13988, I introduced a second commit (0cdb546) making an exception for the `bcrypt` encoder, and always set the `empty-salt` option with it.
We're aware that's not ideal from a OO design perspective, but far better from a DX one. If not desired, I will revert it. 
Anyway I think https://github.com/symfony/symfony/issues/13988#issuecomment-85068382 and other comments about the `PasswordEncoder` API and salt generation should be considered for 3.0.  

A note is added when bcrypt encoder is detected without the `empty-salt` option:
`security:encode-password password "Custom\Class\Bcrypt\User" -n`
screenshot 2015-03-30 a 19 46 36
# Default user-class

The default `user-class` is set to `Symfony\Component\Security\Core\User\User`.
I think this makes sense, as in the previous version of the command, the Symfony\Component\Security\Core\User\User was configured in the setAutocompleterValues and set the $value to Symfony\Component\Security\Core\User\User if null.

Asking the question to the user with the interactive command would have been useful only if the user was able to pick one encoder from a choice list.
# Arguments order / make arguments options

When we had both `salt` and `user-class` as arguments, the command looked like:

``` sh
security:encode-password [password] [user-class] [salt]
```

All arguments were optional, for the interactive command purpose (The command asked the user for missing arguments).
But, as they were arguments, we had to provide each of them in the proper order. That meant we couldn't provide a salt without defining the optional user-class.
So I suggested using options instead of arguments for both user-class & salt.

But as the `salt` option/argument is removed, now I don't feel the need for the `user-class` to be an option.
Indeed, the new command short version will look like the following:

``` sh
#Default user-class: Symfony\Component\Security\Core\User\User
security:encode-password password

#Another user-class:
security:encode-password password "AppBundle\Model\User"
```

Making the user-class an option IMO isn't a necessity anymore, and will only lengthen the command:

``` sh
security:encode-password password --user-class="AppBundle\Model\User"
```
## Bonus:
- [The new command documentation](https://cloud.githubusercontent.com/assets/2211145/6845201/48a66382-d3b2-11e4-8227-b799215a2783.PNG). Thanks to @javiereguiluz.
- [Full interactivity output](https://cloud.githubusercontent.com/assets/2211145/6906381/d7753ce4-d72e-11e4-8547-2ef35c6257e9.PNG)

### CODE REVIEW

**Comment by ogizanagi** [comment]
@javiereguiluz : May I ask you what you think, at least about the new command output (according to command style guide) ?
Is this relevant enough given the #14017 PR and what we debated in #13988 ?

Ping @inanimatt :)

**Comment by javiereguiluz** [comment]
@ogizanagi honestly I think we should change everything. The output is very different from what we propose for Symfony commands. Sadly, I also find the text contents displayed by the command very verbose :(

**Comment by ogizanagi** [comment]
@javiereguiluz : Actually, I think you're right. I believe the command was meant to be used by symfony's newcomers for didactic purposes. Which explains the initial full interactivity and verbosity.
Even getting rid of *useless* questions, I only made the output heavier by adding those notes and salt generation confirmation... The help message itself is a big piece...
Maybe it should only be a very simple utility.

Should we keep, remove, or tweak:
- In interactive mode:
    - salt generation confirmation
    - salt generation notice
- any mode:
    - command introduction
    - bcrypt exception and notice
- In the help:
    - security.encoders example
    - salt generation explanations
    - non interactive description

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 1]
```diff
@@ -32,10 +35,10 @@ protected function configure()
     {
         $this
             ->setName('security:encode-password')
-            ->setDescription('Encode a password.')
+            ->setDescription('Encodes a password.')
             ->addArgument('password', InputArgument::OPTIONAL, 'Enter a password')
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: We could chnage this description by something like:  `The raw password to encode`

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 2]
```diff
@@ -32,10 +35,10 @@ protected function configure()
...
-            ->addArgument('user-class', InputArgument::OPTIONAL, 'Enter the user class configured to find the encoder you need.')
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: And this description could be: `The User entity class path associated with the encoder used to encode the password`

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 3]
```diff
@@ -32,10 +35,10 @@ protected function configure()
...
-            ->addArgument('salt', InputArgument::OPTIONAL, 'Enter the salt you want to use to encode your password.')
+            ->addArgument('user-class', InputArgument::OPTIONAL, 'Enter the user class configured to find the encoder you need.', 'Symfony\Component\Security\Core\User\User')
+            ->addOption('empty-salt', null, InputOption::VALUE_NONE, 'Do not generate a salt or let the encoder generate one.')
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: This option oculd be renamed to `generate-salt` and be `false` by default. And the description could be: `Generate a random salt and add it to the encoded password`

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 4]
```diff
@@ -75,29 +82,38 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
         $password = $input->getArgument('password');
-        $salt = $input->getArgument('salt');
         $userClass = $input->getArgument('user-class');
+        $emptySalt = $input->getOption('empty-salt');
 
         $helper = $this->getHelper('question');
 
         if (!$password) {
+            if (!$input->isInteractive()) {
+                throw new \Exception('The password must not be empty.');
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: Why not? I know that it's an edge case, but someone would like to encode a blank password (using some random salt).

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 5]
```diff
@@ -32,10 +35,10 @@ protected function configure()
...
-            ->setDescription('Encode a password.')
+            ->setDescription('Encodes a password.')
             ->addArgument('password', InputArgument::OPTIONAL, 'Enter a password')
-            ->addArgument('user-class', InputArgument::OPTIONAL, 'Enter the user class configured to find the encoder you need.')
-            ->addArgument('salt', InputArgument::OPTIONAL, 'Enter the salt you want to use to encode your password.')
+            ->addArgument('user-class', InputArgument::OPTIONAL, 'Enter the user class configured to find the encoder you need.', 'Symfony\Component\Security\Core\User\User')
+            ->addOption('empty-salt', null, InputOption::VALUE_NONE, 'Do not generate a salt or let the encoder generate one.')
             ->setHelp(<<<EOF
 
 The <info>%command.name%</info> command allows to encode a password using encoders
```
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: I find the command help very confusing. Here it is a proposal for changing it:

```
The <info>%command.name%</info> command encodes passwords according to your
security configuration. This command is mainly used to generate passwords for
the <comment>in_memory</comment> user provider type and for changing passwords
in the database while developing the application.
 
Suppose that you have the following security configuration in your application:

<comment>
# app/config/security.yml
security:
    encoders:
        Symfony\Component\Security\Core\User\User: plaintext
        AppBundle\Entity\User: bcrypt
</comment>

If you execute the command non-interactively, the default Symfony User class
is used and a random salt is generated to encode the passwod:

  <info>php %command.full_name% --no-interactive [password]</info>

Pass the full user class path as the second argument to encode passwords for
your own entities:

  <info>php %command.full_name% --no-interactive [password] AppBundle\Entity\User</info>

Executing the command interactively allows you to provide a random salt for
encoding the password:

  <info>php %command.full_name% [password] AppBundle\Entity\User</info>

In case your encoder doesn't require a salt, add the <comment>empty-salt</comment> option:

  <info>php %command.full_name% --empty-salt [password] AppBundle\Entity\User</info> 

```

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php (hunk 3 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: My main concern of this command from a DX point of view is the management of the random salt. This is the behavior that I'd like to see:

  * Assume that users never want to provide the salt value explicitly.
  * Therefore, by default, the command manages the salt:
    * If it's Bcrypt, don't generate a salt.
    * If it's an encoder of any other type, generate a salt.
  * If the users want to use their own salts, allow to do that via an option (`--salt` ?)

So this would be the behavior of the command:

```bash
// encodes the password for the default Symfony User class and generates the salt randomly
// if we use bcrypt, the command doesn't generate a salt, but the user is not informed about that
$ php app/console security:encode-password PASSWORD

// encodes the password for our custom User class and generates the salt randomly
// if we use bcrypt, the command doesn't generate a salt, but the user is not informed about that
$ php app/console security:encode-password PASSWORD AppBundle\Entity\User

// encodes the password for the default Symfony User class and uses our custom salt
// if we use bcrypt, the command doesn't use this salt and it informs about it to the user
$ php app/console security:encode-password PASSWORD --salt="..."

// encodes the password for our custom User class and uses our custom salt
// if we use bcrypt, the command doesn't use this salt and it informs about it to the user
$ php app/console security:encode-password PASSWORD AppBundle\Entity\User --salt="..."
```

Comment (inanimatt) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: @javiereguiluz I agree very strongly with the first two bullets, but not with the last one. I really don't see the value of being able to specify a salt, so even giving the option makes the command more complicated and the behaviour more unexpected (why accept a salt option and then discard it if encoding with bcrypt? You can totally do it, but it's almost always a really crappy idea - and that's equally true for *all* salt-using encoders, see what I mean? The only difference is the bcrypt encoder is smart enough to make using a cryptographically strong generated salt the default behaviour).

I think good DX is about making the right thing easy, and making mistakes more difficult; specifying a salt is almost always a mistake. For those instances where it's *not* a mistake, I think that's a sufficiently power-user situation that this command has no obligation to cater to it :stuck_out_tongue:

So IMO `--salt` should be removed. Ideally `--empty-salt` should also be removed *eventually*, but is probably necessary for some custom encoders until a PR fixes the leaky abstraction in the password encoder design.

Comment (jrobeson) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: quoting http://blog.ircmaxell.com/2015/03/thoughts-on-design-of-apis.html

This technique is definitely not perfect. It's quite hard to get the tradeoff line right. For example, I think I made a big mistake with password_hash() allowing the user to specify a salt manually. In all (literally all) the cases I've seen users specifying a salt manually, it's either of a lower quality than would be generated automatically, or it's worse than that.

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: If we add the `--salt` option I don't think the command necessarily complicates:

  * If you know nothing about the salt, you won't use this option and the command won't ask you for it.
  * If you know what a salt is, but don't use this option, the command won't bother you and everything will work.
  * If you are an advanced user, read the command help, and nedd (for whatever reason) to use a custom salt, you just use this super easy `--salt` option. There is only one exception: if you are also using Bcrypt, the command will remind you that you cannot provide a salt in this case, will ignore it and everything will work as expected.

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: @jrobeson it's true that no sane developer would provide a salt manually, but one of the premises of Symfony is: *"never assume anything on behalf of the developer"*. I find it weird, but legitimate, the need to provide a custom salt in some cases.

Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: What is the legitimate use case?

This command should help users for the most common use cases, and for those, you never have to provide a salt manually.

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: I can think of at least of two cases, but I don't know if they are legitimate.

This is the code used by this command to generate the salt:

```php
private function generateSalt()
{
    return base64_encode($this->getContainer()->get('security.secure_random')->nextBytes(30));
}
```

  * First case: you are using a database that cannot be changed and where the `salt` table column is 10 bytes long.
  * Second case: your company security rules enforce 100-bytes salt values.

Comment (jrobeson) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: if people need to do that, they can override the command and provide their own implementation of generateSalt.  Providing a custom salt is the anti DX imo.

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: @jrobeson are you sure that the `--salt` option is anti DX?

**Scenario:** your database requires 10 bytes salt values.

**Proposal 1**:

Execute: `php app/console security:encode-password raw_password --salt="ae037d6f72"`

**Proposal 2**:

1. Create a new Symfony console command
2. Make this command extend the default SecurityEncodePasswor command
3. Override the `generateSalt()` command and think of a better implementation
4. Execute `php app/console security:encode-password raw_password`

Comment (jrobeson) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: and how often will that actually happen?

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: I don't know. But let's go back to the original discussion. What I propose is the following:

  * Don't require to provide a salt and never bother the user about the salt. The command does always the right thing automatically: generate it randomly or not generate it if using Bcrypt.
  * If the user, for whatever reason, needs to provide a custom salt, let him/her use the `--salt` option.

Comment (inanimatt) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: Thank you @fabpot 

@javiereguiluz "never assume anything on behalf of the developer" is a guideline. It's a **great** guideline. But it's not dogma. I don't think the meaning of the guideline is "we will let you shoot yourself in the foot if you want because theoretically you might _need_ to shoot yourself in the foot", it's more like "if we assume too much we risk making your job harder not easier" - this is not one of those instances. The only legitimate use-case I can think of for providing a known salt to the encoder is to verify its output (i.e. to test that it works), but that's not the job of this command.

I don't think your use-cases are legitimate, sorry. Frankly I'm not even convinced that _my_ use case is legitimate. They're theoretically plausible, but the solution to them isn't to make this command generate the salt, it's to make a custom PasswordEncoderInterface implementation that generates conforming salts in the same way that the BCrypt one does.

Basically this whole argument has come about because the design of that component is flawed - it's a leaky abstraction. Encoders right now demand that we have special knowledge of the length and composition of the salt, which is wrong. We don't know and can't know if the encoder accepts raw binary salts or base64 or (god help us) only hexits; we don't know how long the salt needs to be. This is wrong. Until that's fixed, I think we should do the best we can, which is to generate a reasonably long base64 salt for everything except BCrypt. Accepting a salt override (or even an empty-salt) is a mistake - it shouldn't be any of our business (in the context of this command), and it's a burden and a problem for users, not a benefit.

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: I guess one of the reasons of this long discussion is that we're thinking about different use cases for this command. I proposed to updated the description of this command as follows:

```
The <info>%command.name%</info> command encodes passwords according to your
security configuration. This command is mainly used to generate passwords for
the <comment>in_memory</comment> user provider type and for changing passwords
in the database while developing the application.
```

You don't use this command in the real application, you only use it for `in_memory` providers and for quick tests.

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php (hunk 4 above)
Comment (ogizanagi) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: Really ? Should we encourage this ? What would be the point of having a password if blank and using this command to generate it (even for development) ? :/

Comment (fabpot) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: I agree with @ogizanagi, we must not allows empty passwords here.

**Comment by ogizanagi** [comment]
Changes have been made. Thanks @javiereguiluz for your precious corrections & samples for the command description and @jrobeson & @inanimatt for having reminded what was said about the `salt` option before.

Should we at least make the `generateSalt` method protected in order to let users extend the command easily ? (Eventually give the encoder instance as first argument, if they have to handle multiple encoders in their application). I'm not really happy with this idea, but...

**Comment by ogizanagi** [comment]
@javiereguiluz : I didn't expect changing the output that much within this PR, only the behavior, but do you think the following command outputs are more compliant with your console style guide ? (I used the recently merged #14057 , but changes haven't been pushed for now)

I'll be glad to change the note about salt generation when using the interactive way if it doesn't please you, but I still think the user should be warn somehow (or explain the confirmation question context).

#### No interaction:

##### No password:
screenshot 2015-03-30 a 19 46 07

##### Bcrypt without empty salt option
screenshot 2015-03-30 a 19 46 36

##### Bcryp with empty-salt option
screenshot 2015-03-30 a 19 47 33

#### Interaction & other encoder
screenshot 2015-03-30 a 19 49 05

**Comment by javiereguiluz** [comment]
@ogizanagi wow! I love it! Now that the new console style guide is merged, I was planning to start updating the existing commands, but I'm glad that you already started with this command. My only suggestion: there is no need to add `✔` when the result is successful because we already provide the `[OK]` prefix and the color background.

**Comment by ogizanagi** [comment]
Changes pushed, then ! I will update the PR description and screenshots.

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php [hunk 6]
```diff
@@ -72,154 +84,86 @@ protected function configure()
...
-            $saltQuestion = $this->createSaltQuestion($input, $output);
-            $salt = $helper->ask($input, $output, $saltQuestion);
-        }
+        $salt = null;
+
+        if ($input->isInteractive() && !$emptySalt) {
+            $emptySalt = true;
 
-        $output->writeln("\n <comment>Encoders are configured by user type in the security.yml file.</comment>");
+            $output->note('The command will take care about generating a salt for you. Be aware that some encoders advise to let them generate their own salt. If you\'re using one of those encoders, please answer \'no\' to the question below. '.PHP_EOL.'Provide the \'empty-salt\' option in order to let the encoder handle the generation itself.');
```
Comment (inanimatt) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: **Very** minor point: "The command will generate a salt for you" would be more simple and more correct (otherwise "take care of") :smile:

**Comment by inanimatt** [comment]
Nice work, @ogizanagi, it's very much improved. And thanks again (and sorry) to @saro0h for the initial work on this. We got there in the end :)

File: src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php (hunk 6 above)
Comment (ogizanagi) [inline src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php]: Thank you. I updated this right now. :)

**Comment by ogizanagi** [comment]
Yes, I feel bad for having change this much @saro0h original work :disappointed_relieved: 
I hope it would be for the best.

**Comment by ogizanagi** [comment]
What should be done about Travis fails in [56604789](https://travis-ci.org/symfony/symfony/jobs/56604789) and [56604790](https://travis-ci.org/symfony/symfony/jobs/56604790) ?
Sorry, I don't fully understand the impact of the `deps` env variable  `low` and `high` values and the 
> Fatal error: Class 'Symfony\Component\Console\Style\SymfonyStyle' not found in /home/travis/build/symfony/symfony/src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php on line 87

 error That travis reported.
If it relies on lowest components, this build doesn't matter for our case, right ?
But then, what's going on with `high` ?

**Comment by ogizanagi** [comment]
ping @fabpot . I think this PR is ready for merging. Let me know if there is anything else to do (squash ?).

**Comment by javiereguiluz** [comment]
:+1: indeed the code looks ready to be merged ... but what should we do with the Travis issues?

**Comment by ogizanagi** [comment]
@javiereguiluz : at least for 3.0 branch, we'll need this PR (https://github.com/symfony/symfony/pull/14220) one day or another.
For `deps=low`, AFAIU, I think we can safely ignore this build.

**Comment by nicolas-grekas** [comment]
In some cases deps=high can be ignored. But deps=low should never ever be red before merging anything.
Well, at this very moment, the upstream is red so you have no choice I know.

**Comment by fabpot** [comment]
Thank you @ogizanagi.

**Comment by nicolas-grekas** [comment]
@ogizanagi tests are failing on 2.7 after the merge, can you please provide a fixing patch asap?

**Comment by ogizanagi** [comment]
@nicolas-grekas : Sure. But I think the error comes from this commit: https://github.com/symfony/symfony/commit/e9c7912f572b70c3704df20dc7008e8bbf0df4ed
